### ISSUE TITLE
브랜치 전략: feature-robust-optim

### ISSUE THREAD

**Issue by iejob** [iejob/computer-vision-competition-1SEN#10]
## 목표와 범위
- **목표:**
    실제 배포 환경(회전·블러·노이즈·압축 손상 등) 및 데이터 편차에 견고한 모델을 위해 **강건 증강·손실/정규화·최적화 레시피**를 표준화합니다.
    
    - 본 브랜치는 **AugMix/RandAugment/Color/Geo 강화**, **SAM/ASAM/AWP/EMA/Label Smoothing/Gradient Clipping** 등 **강건화 기법의 조합 설계·튜닝·검증·리포팅**을 담당합니다.
    - 모든 기능은 **`ipynb`에서 사전 검증** 후, 통과 코드만 `src/`의 표준 디렉터리로 **모듈화 이관**합니다.
    - 병합 전 **강건성 지표(부가 코럽션 벤치) 개선, CV 유지·개선, 재현성**을 완료 기준(DoD)으로 정의합니다.
- **범위:**
    - **증강 정책**: 문서 이미지 특화 **기하/포토메트릭** 조합(AugMix, RandAugment, CutMix/MixUp 호환 설계).
    - **최적화/정규화**: **SAM/ASAM/EMA/Label Smoothing/Focal/Gradient Clipping/Weight Decay/TTA 정책**.
    - **학습 스케줄**: **Cosine+Warmup**, Progressive Resizing(선택), Accumulation/AMP 연계.
    - **강건성 평가**: **Validation 코럽션 스위트**(회전/블러/노이즈/압축/저조도 등)에서 Macro-F1 평균(=Robustness Score) 산출.
    - **리포팅**: 강건성 지표·트레이드오프 표·권장 레시피 문서화.
    - **규칙 준수**: 외부 데이터/유료 API/테스트 데이터 활용 금지(PL 금지).

---

### 1) 증강 정책(Train/TTA)
- **Train 강건 증강(추가)**
    - **RandAugment**(N=2~~3, M=7~~11), **AugMix**(mixture width=3, depth=1~2, α=1.0) 선택형.
    - **기하**: RandomRotation(±20~~30°), RandomAffine(shear/scale 소폭), Perspective(약), RandomResizedCrop(scale 0.8~~1.0).
    - **포토메트릭**: ColorJitter(0.1~~0.3), CLAHE(≤0.3), GaussianBlur(커널 3~~5), JPEGCompression(quality 60~90), MotionBlur(약).
    - **라벨 믹싱**: CutMix/MixUp(α=0.2~0.4) 옵션 **호환 모드**(손실 스케줄과 충돌 방지).
    - **가드레일**: 문서 구조(텍스트/레이아웃) 훼손 방지 – 강도 상한 및 조합 제약(표로 명세, 아래 “파라미터 표준”).
- **TTA(추론)**
    - **회전 축**: {0°, 90°, 180°, 270°}, **Flip**: {None, H} 조합 최소셋.
    - **색상 약변환**: 단 1종 선택(예: 밝기 ±5%) – 과도한 포토메트릭 TTA 금지.
    - **집계**: **logit 평균**(권장) 또는 확률 평균.

### 2) 손실·정규화
- **Label Smoothing**: ε=0.1(권장; MixUp/ CutMix와 중복 과다 사용 금지 → 동시 사용 시 ε=0.05로 자동 완화).
- **Focal Loss**: γ=2(선택; 클래스 불균형·하드 샘플 대응) – **Label Smoothing과 동시 사용 비권장**.
- **Weight Decay**: 1e-4~5e-4 범, 백본·배치·이미지 크기에 따라 스윕.
- **Stochastic Depth**(Transformer류 사용 시): p=0.1 이하(작게).

### 3) 강건 최적화
- **SAM/ASAM**: ρ(perturb radius)=0.05~0.2 스윕, base optimizer=AdamW.
- **AWP(Adversarial Weight Perturb.)**: ε_w=0.005~0.02(선택), warmup 이후 활성화.
- **EMA**: decay=0.999(에폭 3 이후 업데이트) – 검증/추론 시 EMA weights 사용.
- **Gradient Clipping**: global-norm ≤ 1.0.
- **Progressive Resizing**: (224→320) 또는 (224→384) 구간적 단계 상승(학습 후반 30~40%부터).
- **AMP + Accumulation**: 메모리 한계 시 `acc_steps`로 유효 배치 보존.

### 4) 스케줄·학습 레시피
- **Scheduler**: CosineAnnealing + Linear Warmup(5%) 기본, OneCycleLR 옵션.
- **권장 레시피(3종)**
    - *Robust-Light*: RandAugment(N=2,M=7) + LS(0.1) + EMA + Cosine, no SAM.
    - *Robust-Core(기본)*: AugMix(depth=2,α=1.0) + LS(0.1) + **SAM(ρ=0.1)** + EMA + Cosine.
    - *Robust-Heavy*: RandAugment(N=3,M=11) + CutMix(α=0.2) + **ASAM(ρ=0.15)** + EMA + Progressive Resizing(224→320).
- **충돌 규칙**
    - Focal ↔ Label Smoothing **동시 비권장**(둘 중 하나).
    - AugMix ↔ 과도한 JPEG/MotionBlur **중첩 금지**(텍스트 훼손 위험).
    - SAM/ASAM ↔ 高 M의 RandAugment **과도 조합 지양**(수렴 불안정) → Warmup 단계에서는 RA 강도 완화.

### 5) 강건성 평가(Validation Corruption Suite)
- **코럽션 타입(문서 도메인)**
    - Geo: SmallRotation(±15°), Perspective(약), MildCrop/Pad(비율 유지)
    - Blur/Noise: GaussianBlur(σ 0.5~~1.5), MotionBlur(약), GaussianNoise(σ 5~~15)
    - Photometric: Brightness(±15%), Contrast(±15%), ColorShift(소폭)
    - Compression: JPEG(quality 70/50)
- **Severity 1~3** 레벨별 생성 → **각 타입×severity에 대한 Macro-F1** 측정 후 평균 = **Robustness Score**.
- **평가지표**
    - Clean Macro-F1(깨끗한 val)
    - Robustness Score(코럽션 평균 Macro-F1)
    - ΔRobustness = Robustness Score − Baseline Score
    - Trade-off(클린 대비 손실 %p)

### 6) 로깅·리포팅
- **표준 로그**: `experiments/robust/<run_id>/{config.json, train_log.csv, robust_eval.csv, summary.json}`
- **그림/표**: `robust_vs_clean_plot.png`, `corruption_breakdown.png`, `ttasweep_plot.png`
- **HTML 리포트**: `docs/robust_report_<model>_<cfg>_<timestamp>.html`
    - 요약(클린/로버스트/Δ), 설정, 코럽션별 표, TTA 효과, 권장 레시피.

---

## ipynb-우선 워크플로우 (파일명·셀 단위 체크리스트)
1. **`notebooks/ROBUST/00_robust_config_and_guardrails.ipynb`**
    - [ ]  `configs/robust_config.json` 작성(증강/최적화/스케줄/TTA/코럽션 스위트 파라미터)
    - [ ]  증강 강도·조합 가드레일 검증(과도 변형 차단)
    - [ ]  레시피 3종(*Light/Core/Heavy*) 프리셋 저장
2. **`notebooks/ROBUST/01_robust_augmentations_gallery.ipynb`**
    - [ ]  AugMix/RandAugment/Geo/Photo 예시 갤러리 생성
    - [ ]  텍스트 가독성 체크(사내 기준표) 통과 여부 기록
3. **`notebooks/ROBUST/02_sam_asam_awp_ablation.ipynb`**
    - [ ]  SAM/ASAM/AWP 유무·강도 ρ/ε_w 스윕(소규모 epoch)
    - [ ]  수렴·안정성(발산/NaN)과 Clean/Robust 지표 비교
4. **`notebooks/ROBUST/03_progressive_resize_and_mixup_cutmix.ipynb`**
    - [ ]  Progressive Resizing on/off, CutMix/MixUp α 스윕
    - [ ]  Label Smoothing과의 상호작용 평가
5. **`notebooks/ROBUST/04_tta_policy_sweep.ipynb`**
    - [ ]  회전/Flip/색상 약변환 조합별 TTA 스윕 → **logit 평균** vs prob 평균 비교
    - [ ]  과도 TTA로 인한 성능 하락/시간 증가 trade-off 표
6. **`notebooks/ROBUST/05_validation_corruption_benchmark.ipynb`**
    - [ ]  코럽션 스위트 생성(유효성 샘플) & Macro-F1 측정 → **Robustness Score** 산출
    - [ ]  Baseline 대비 ΔRobustness 요약
7. **`notebooks/ROBUST/06_export_robust_modules.ipynb`**
    - [ ]  통과 코드만 `src/training/*`, `src/utils/*`로 이관
    - [ ]  `__all__`/docstring/타입힌트/사용 예제 정리

---

## 산출물(Artifacts)
- **코드 모듈**
    - `src/training/augment/policies.py` : RandAugment/AugMix/문서 특화 정책
    - `src/training/augment/tta.py` : TTA 프로토콜(회전/flip/약변환, logit 평균)
    - `src/training/losses.py` : CE(+Label Smoothing), Focal, MixUp/CutMix 보조 유틸
    - `src/training/optimizers/sam.py` : SAM/ASAM 스텝 구현
    - `src/training/optimizers/awp.py` : Adversarial Weight Perturbation
    - `src/training/schedulers.py` : Cosine+Warmup/OneCycle 팩토리
    - `src/training/regularizers.py` : EMA, Grad Clipping, Stochastic Depth 헬퍼
    - (기존 연계) `src/training/train.py` : 훅 기반으로 상기 모듈 플러그형 연동
- **리포트/로그**
    - `docs/robust_report_<model>_<cfg>_<timestamp>.html`
    - `experiments/robust/<run_id>/{config.json, train_log.csv, robust_eval.csv, robust_vs_clean_plot.png, corruption_breakdown.png, summary.json}`
- **설정 파일**
    - `configs/robust_config.json`
- **노트북(최종 파일명 확정)**
    - `notebooks/ROBUST/00_robust_config_and_guardrails.ipynb`
    - `notebooks/ROBUST/01_robust_augmentations_gallery.ipynb`
    - `notebooks/ROBUST/02_sam_asam_awp_ablation.ipynb`
    - `notebooks/ROBUST/03_progressive_resize_and_mixup_cutmix.ipynb`
    - `notebooks/ROBUST/04_tta_policy_sweep.ipynb`
    - `notebooks/ROBUST/05_validation_corruption_benchmark.ipynb`
    - `notebooks/ROBUST/06_export_robust_modules.ipynb`

---

### 1 증강·TTA
- **Train**: 정책 선택형(`randaugment`/`augmix`/`custom_geo_photo`)과 강도 범위 설정.
- **TTA**: 회전·Flip 최소셋 + 색상 약변환 1종, **logit 평균** 지원.
- **가드레일**: 텍스트 가독성 임계 도달 시(예: OCR 유사 지표 하락) 해당 조합 자동 비활성화(옵션).

### 2 최적화·정규화
- **SAM/ASAM/AWP**: on/off 및 강도 파라미터(ρ, ε_w) 설정, Warmup 이후 단계적 활성화.
- **EMA**: 학습 3epoch 이후 업데이트, 검증·추론 시 EMA 가중치 사용.
- **Loss**: CE(+LS) 또는 Focal 중 택1, MixUp/CutMix 동시 사용 시 LS 자동 감쇠.

### 3 스케줄·학습
- **Cosine+Warmup** 기본, Progressive Resizing(후반부 224→320/384) 옵션.
- **AMP/Accumulation** 지원, imgs/sec 로깅.

### 4 강건성 평가
- **코럽션 스위트**: 타입×severity=최소 8×3 조합 → Macro-F1 평균(**Robustness Score**) 출력.
- **클린-로버스트** 동시 보고, Trade-off 표 생성.

### 5 로깅·리포트

- 모든 설정·결과를 JSON/CSV/PNG로 저장, HTML 리포트에 자동 포함.

---

## 실험·검증(Validation) 기준
| 항목 | 측정/검증 방법 | 통과 기준(제안) |
| --- | --- | --- |
| **클린 성능 유지** | clean val Macro-F1 | 베이스라인 대비 **-0.5%p 이내 손실** 또는 개선 |
| **강건성 개선** | Robustness Score | 베이스라인 대비 **+1.0%p 이상** (초기 목표) |
| **수렴 안정성** | NaN/발산 감지 | 0건, 재시작 없이 종료 |
| **레시피 일관성** | Light/Core/Heavy 비교 | Core가 Light 대비 강건성 **≥ +0.5%p** |
| **TTA 효과** | TTA on/off | on 시 클린 **+0.3~1.0%p** 범 향상, 과적합 징후 없음 |
| **재현성** | 동일 설정 2회 | Robustness Score 편차 **≤ ±0.5%p** |
| **리포트 완결성** | HTML 섹션/그림/표 | 필수 섹션 전부 포함, 링크 정상 |
---

## 병합 조건(Definition of Done & PR 체크리스트)
- **DoD**
    - ✅ 00~06 노트북 **모든 셀 실행 완료(오류 0, 출력 포함)**
    - ✅ `docs/robust_report_*.html` 1부 생성(요약·그림·표 포함)
    - ✅ **Robustness Score** 목표 충족(+1.0%p↑) **및** 클린 성능 손실 ≤0.5%p
    - ✅ 아티팩트(`experiments/robust/<run_id>/…`) 누락 없음
    - ✅ 규칙 준수(외부 데이터/PL/유료 API 미사용)
- **PR 체크리스트**
    - [ ]  코드 포맷팅(black/isort), 린트 통과
    - [ ]  단위 테스트 4종 통과(증강 파이프·SAM/EMA 동작·TTA 집계·코럽션 지표)
    - [ ]  `configs/robust_config.json` 커밋 및 사용 예시 제시
    - [ ]  리포트/로그 링크 및 요약 스크린샷 첨부

---

## 브랜치 작업 순서(권장 커밋 플로우)
1. `feat(config)`: robust_config.json, 가드레일 로직
2. `feat(augment)`: [policies.py](http://policies.py/)(AugMix/RandAugment/문서 전용) + 갤러리 노트북
3. `feat(optim)`: SAM/ASAM/AWP 구현 및 훅 연결
4. `feat(regularize)`: EMA/Label Smoothing/CutMix/MixUp 헬퍼
5. `feat(scheduler)`: Cosine+Warmup/OneCycle 팩토리
6. `feat(tta)`: [tta.py](http://tta.py/)(logit 평균 포함)
7. `feat(bench)`: 코럽션 스위트 및 Robustness Score 산출
8. `feat(ablation)`: ablation 노트북(02/03/04)
9. `feat(report)`: robust_report 생성기
10. `feat(export)`: 모듈 이관(06)
11. `tests/*`: 단위 테스트 추가

---

## 구성·설정 표준
**`configs/robust_config.json` (예시)**

```json
{
  "recipe": "core",
  "img_size": 224,
  "augment": {
    "use_randaugment": false,
    "ra_N": 2,
    "ra_M": 7,
    "use_augmix": true,
    "augmix": { "width": 3, "depth": 2, "alpha": 1.0 },
    "geo": { "rotate_deg": 25, "affine_shear": 5, "perspective": 0.05 },
    "photo": { "jitter": 0.2, "clahe_p": 0.2, "blur_p": 0.2, "jpeg_q": 80 },
    "mixing": { "mixup_alpha": 0.2, "cutmix_alpha": 0.0 }
  },
  "optimizer": {
    "base": "adamw",
    "lr": 0.0003,
    "weight_decay": 0.0001,
    "sam": { "enabled": true, "rho": 0.1, "asam": false },
    "awp": { "enabled": false, "eps": 0.01, "warmup_epochs": 3 }
  },
  "regularization": {
    "label_smoothing": 0.1,
    "focal_gamma": null,
    "ema": { "enabled": true, "decay": 0.999 },
    "grad_clip_norm": 1.0,
    "stochastic_depth": 0.0
  },
  "schedule": {
    "name": "cosine",
    "epochs": 25,
    "warmup_pct": 0.05,
    "progressive_resize": { "enabled": true, "milestone_pct": 0.6, "to_size": 320 }
  },
  "tta": { "rotations": [0, 90, 180, 270], "hflip": true, "light_color_jitter": false, "aggregate": "logit_mean" },
  "corruptions": {
    "types": ["rotation","perspective","gaussian_blur","motion_blur","gaussian_noise","brightness","contrast","jpeg"],
    "severity": [1,2,3]
  },
  "paths": {
    "train_csv": "data/raw/train.csv",
    "img_root": "data/raw/train",
    "folds_csv": "data/processed/folds_k5_seed42.csv",
    "exp_dir": "experiments/robust"
  },
  "seed": 42
}

```

---
## 파라미터 표준(요약)
| 항목 | 기본값 | 허용 범위 | 비고 |
| --- | --- | --- | --- |
| RandAugment N/M | 2/7 | N∈[2,3], M∈[7,11] | 과도 강화 금지 |
| AugMix width/depth/α | 3/2/1.0 | depth ∈ [1,3] | 텍스트 가독성 유지 |
| Rotate | ±25° | ±15~±30° | 문서 도메인 권장 |
| JPEG quality | 80 | 60~90 | Train 강도 상한 |
| Label Smoothing | 0.1 | 0.05~0.2 | MixUp 병행 시 0.05 |
| Focal γ | — | 1~2 | LS와 병행 지양 |
| SAM ρ | 0.1 | 0.05~0.2 | ASAM 시 ρ↑ 가능 |
| AWP ε_w | — | 0.005~0.02 | warmup 후 적용 |
| EMA decay | 0.999 | 0.995~0.9995 | ep≥3부터 |
| Grad Clip | 1.0 | 0.5~1.0 | global-norm |
| Progressive Resize | off | 224→320/384 | 후반부 적용 |
---

## 실패 정책(Failure Policy)
- **수렴 실패/NaN**: 즉시 중단, 최근 체크포인트로 롤백, 강도 자동 감쇠(예: RA M-2, ρ-0.02).
- **가독성 훼손**: 갤러리 기준치 미달 시 해당 조합 비활성화.
- **성능 미달**: Robustness Score +1.0%p 미달 또는 클린 손실 >0.5%p → 레시피 재탐색(강도/ρ/ε_w 조정).
- **아티팩트 누락**: 리포트/로그/요약 파일 중 누락 시 머지 불가.

---

## 리스크·준수사항
- **규칙 준수:** 외부 데이터·유료 API·테스트 데이터 활용 금지(PL 금지).
- **과도 변형 리스크:** 텍스트/레이아웃 훼손 예방을 위한 강도 상한·가드레일 필수.
- **학습 불안정 리스크:** SAM/ASAM/AWP 과도 시 발산 → Warmup·강도 감쇠·스케줄 조정.
- **리소스 리스크:** Progressive Resizing·TTA로 시간/메모리 증가 → AMP/Accumulation·TTA 최소셋 유지.
